File: abstractions/java/lib/src/main/java/com/microsoft/kiota/authentication/BaseBearerTokenAuthenticationProvider.java
package com.microsoft.kiota.authentication;

import com.microsoft.kiota.RequestInformation;

import java.lang.UnsupportedOperationException;
import java.net.URI;
import java.net.URISyntaxException;
import java.util.concurrent.CompletableFuture;
import java.util.Objects;

import javax.annotation.Nonnull;

/** Provides a base class for implementing AuthenticationProvider for Bearer token scheme. */
public class BaseBearerTokenAuthenticationProvider implements AuthenticationProvider {
    public BaseBearerTokenAuthenticationProvider(@Nonnull final AccessTokenProvider accessTokenProvider) {
        this.accessTokenProvider = Objects.requireNonNull(accessTokenProvider);
    }
    private final AccessTokenProvider accessTokenProvider;
    private final static String authorizationHeaderKey = "Authorization";
    public CompletableFuture<Void> authenticateRequest(final RequestInformation request) {
        Objects.requireNonNull(request);
        if(!request.headers.keySet().contains(authorizationHeaderKey)) {
            final URI targetUri;
            try {
                targetUri = request.getUri();
            } catch (URISyntaxException e) {
                return CompletableFuture.failedFuture(e);
            }
            return this.accessTokenProvider.getAuthorizationToken(targetUri)
                .thenApply(token -> {
                    if(token == null || token.isEmpty()) {
                        throw new UnsupportedOperationException("Could not get an authorization token", null);
                    }
                    request.headers.put(authorizationHeaderKey, "Bearer " + token);
                    return null;
                });
        } else {
            return CompletableFuture.completedFuture(null);
        }
    }
}

File: authentication/ruby/base_bearer_token_authentication_provider.rb
# frozen_string_literal: true

require 'concurrent'
require_relative './authentication_provider'
require_relative './access_token_provider'

module MicrosoftKiotaAbstractions
  # Provides a base class for implementing AuthenticationProvider for Bearer token scheme
  class BaseBearerTokenAuthenticationProvider
    include Concurrent::Async

    def initialize(access_token_provider)
      raise StandardError, 'access_token_provider parameter cannot be nil' if access_token_provider.nil?

      @access_token_provider = access_token_provider
    end 

    AUTHORIZATION_HEADER_KEY = 'Authorization'
    def authenticate_request(request, additional_properties)
      raise StandardError, 'Request cannot be null' if request.nil?
      return if request.headers.key?(AUTHORIZATION_HEADER_KEY)

      token = @access_token_provider.get_authorization_token(request, additional_properties)
      raise StandardError, 'Could not get an authorization token' if token.nil?

      request.headers[AUTHORIZATION_HEADER_KEY] = "Bearer #{token}"
    end

  end
end
